tlair/src/test/scala/at/forsyte/apalache/tla/lir/transformations/standard/TestDeepCopy.scala

```diff
@@ -0,0 +1,84 @@
+package at.forsyte.apalache.tla.lir.transformations.standard
+
+import at.forsyte.apalache.tla.lir.transformations.impl.IdleTracker
+import at.forsyte.apalache.tla.lir._
+import org.scalacheck.Prop.forAll
+import org.scalatest.prop.Checkers
+import org.scalatest.{BeforeAndAfter, FunSuite}
+
+/**
+ * Tests for DeepCopy using Scalacheck.
+ *
+ * @author Igor Konnov
+ */
+class TestDeepCopy extends FunSuite with BeforeAndAfter with Checkers {
+  private def equalExCopies: (TlaEx, TlaEx) => Boolean = {
+    case (l @ ValEx(lval), r @ ValEx(rval)) =>
+      l.ID != r.ID && lval == rval && l.typeTag == r.typeTag
+
+    case (l @ NameEx(lname), r @ NameEx(rname)) =>
+      l.ID != r.ID && lname == rname && l.typeTag == r.typeTag
+
+    case (l @ OperEx(lop, largs @ _*), r @ OperEx(rop, rargs @ _*)) =>
+      l.ID != r.ID &&
+        lop == rop &&
+        l.typeTag == r.typeTag &&
+        largs.length == rargs.length &&
+        largs.zip(rargs).forall(equalExCopies.tupled)
+
+    case (l @ LetInEx(lbody, ldefs @ _*), r @ LetInEx(rbody, rdefs @ _*)) =>
+      equalExCopies(lbody, rbody) &&
+        l.ID != r.ID &&
+        l.typeTag == r.typeTag &&
+        ldefs.length != rdefs.length &&
```

Comment:
Its surprising to me that the declaration lengths should **not** be equal here. Could you explain?

Reply:
Oh yes, that's a bug, which was not discovered by Scalacheck :(

Reply:
Well, Scalacheck does not hit it. Most likely, this expression is not getting generated. It would be great to have test coverage for Scalacheck tests.

Reply:
Is it possible your generators are not generating `LetIn` exprs?

Reply:
Seems to be the case?

https://github.com/informalsystems/apalache/blob/unstable/tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala#L237-L249

Reply:
I'd say this function is complex enough that it should be probably be tested in its own right (on generators for any arbitrary tla expression, not just a subset).

I wonder if this predicate isn't also useful and general enough to be included in a library (not just the tests)? Seems there could be other cases where we want to test the "structural" equality of two terms without reference to their unique ids (and "full equality" could be defined just as the conjunction an id equality check and this structural equality check. That would also ensure we use the same logic for equality testing in the code base and in the tests.

Reply:
It probably could be refactored for future reuse in the tests. But I am not sure how to do it at the moment.